src/main/scala/rise/Cuda/primitives.scala

```diff
@@ -0,0 +1,92 @@
+package rise.Cuda
+
+import rise.core.{Builder, Primitive}
+import rise.macros.Primitive.primitive
+import rise.core.DSL.Type._
+import rise.core.types._
+
+object primitives {
+
+  protected def mapType: Type =
+    impl{n: Nat =>
+      impl{s: DataType =>
+        impl{t: DataType => (s ->: t) ->: ArrayType(n, s) ->: ArrayType(n, t)}}}
+
+  @primitive case class mapBlock(dim: Char) extends Primitive with Builder {
+    mapType
+  }
+
+  @primitive case class mapGlobal(dim: Char) extends Primitive with Builder {
+    mapType
+  }
+
+  @primitive case class mapThreads(dim: Char) extends Primitive with Builder {
+    mapType
+  }
+
+  @primitive case class mapWarp(dim: Char) extends Primitive with Builder {
+    mapType
+  }
+
+  @primitive case class mapLane(dim: Char) extends Primitive with Builder {
+    mapType
+  }
+
+  @primitive object asFragment extends Primitive with Builder {
+    impl{rows: Nat =>
+      impl{columns: Nat =>
+        impl{d3: Nat =>
+          impl{dt: DataType =>
+            impl{fragType: FragmentKind =>
+              impl{matrixLayout: MatrixLayout =>
+                ArrayType(rows, ArrayType(columns, dt)) ->: FragmentType(rows, columns, d3, dt, fragType, matrixLayout)}}}}}}
+  }
+
+  @primitive object asMatrix extends Primitive with Builder {
+    impl{rows: Nat =>
+      impl{columns: Nat =>
+        impl{d3: Nat =>
+          impl{dt: DataType =>
+            FragmentType(rows, columns, d3, dt) ->: ArrayType(rows, ArrayType(columns, dt))}}}}
+  }
+
+  @primitive object generateFragment extends Primitive with Builder {
+    impl{rows: Nat =>
+      impl{columns: Nat =>
+        impl{d3: Nat =>
+          impl{dt: DataType =>
+            impl{fragType: FragmentKind =>
+              dt ->: FragmentType(rows, columns, d3, dt, fragType, MatrixLayout.Row_Major)}}}}}
+  }
+
+  @primitive object tensorMMA extends Primitive with Builder {
+    impl{layoutA: MatrixLayout =>
+      impl{layoutB: MatrixLayout =>
+        impl{m: Nat =>
+          impl{n: Nat =>
+            impl{k: Nat =>
+              impl{dt: DataType =>
+                impl{dt2: DataType =>
+                  FragmentType(m, k, n, dt, FragmentKind.AMatrix, layoutA) ->:
+                    FragmentType(k, n, m, dt, FragmentKind.BMatrix, layoutB) ->:
+                    FragmentType(m, n, k, dt2) ->: FragmentType(m, n, k, dt2)}}}}}}}
+  }
+
+  @primitive object toSharedMemoryShift extends Primitive with Builder {
+    impl{m: Nat =>
+      impl{n: Nat =>
+        expl((_: Nat) =>
+          impl{t: DataType =>
+            ArrayType(m, ArrayType(n, t)) ->: ArrayType(m, ArrayType(n, t))})}}
+  }
+
+  @primitive object mapFragmentElements extends Primitive with Builder {
+    impl{dt: DataType =>
+      impl{fragType: DataType =>
+          (dt ->: dt) ->: fragType ->: fragType}}
+  }
+
+  @primitive object globalToShared extends Primitive with Builder {
```

Comment:
This primitiv copies data from global memory to shared memory. It can only be used if data are currently in global memory (not in private memory).
Using the `toLocal` / `oclToMem` generates the following code:

    ```
    for ( i = ... ) {​
        // uses intermediate register
        // {int tmp=g[i]; smem[i]=tmp;}
        smem[i] = gldata[i]; ​
    }
    ```
    
In CUDA is more efficiently (for compute capability >= 8) to copy like this:    

    ```
    pipeline pipe;
    // load element into shared mem
    for ( i = ... ) {​
        // initiate async memory copy
        memcpy_async(smem[i], 
                    gldata[i], 
                    pipe); ​
    }​

    // wait for async-copy to complete
    pipe.commit_and_wait();​
    ```
Using the memcpy_async there are no intermediate register required. The position of the `pipe.wait()` could be optimized (which is currently not done).
This method to copy data is implemented with the `globalToShared` primitive.
Is it may possible redefine the code generation for the `oclToMem` primitive when using CUDA and copy data from global memory to shared memory?

Reply:
Ok. @LukasSiefke thanks for the explanation. Can you add a TODO to this primitive so that we later can investigate how we can make `oclToMem(Local)` (or the equivalent future CUDA version) generate this type of code.